Update Schema: Prompt becomes a table and add subprompt table

andrew-rietz opened the pull request:
- Prompt has become its own table (previously in flashcard)
- Added sub-prompt table

Issues:
- No option to create a user-defined type in the DBDesigner.net interface. Currently using json as the type for both the "prompt" and "sub_prompt" fields within the [prompt] and [sub-prompt] tables respectively.

Addresses #17 and #18

andrew-rietz commented:
Taylor, it looks like my solution for generating the ERD and schema file may not work out after all. Kind of a bummer... Take a look and let's circle back. I wasn't able to create a user-defined type ("multiple_choice") but opted to use a json data type instead. I think it's workable, but may require rethinking how we handle input and database I/O.

tmrietz commented on schema/schema_updated_17.sql:
```diff
@@ -0,0 +1,88 @@
+
+CREATE TABLE "prompt" (
+	"id" serial NOT NULL DEFAULT 'uuid_generate_v1()',
+	"type" json NOT NULL,
```
I think we can make "type" a varchar(30) type and add a check that enforces 2 types right now: "text" and "multiple_choice"

andrew-rietz replied:
I get it now... I thought you wanted some type of custom data type. Really you want the text values within "type" to be either "text" or "multiple_choice".


This should be updated and incorporated in latest commit

tmrietz commented:
Can we limit the whitespace a bit? And do we need to include the binary file? Other than that and comment on the "type" field's type, looks good.

andrew-rietz commented:
Yes, we can limit the whitespace. I had been using the DBDesigner.net site to create the ERD and automatically generate the underlying sql code, and didn't do any cleaning after exporting. It sounds like we have needs that are beyond the scope of what DBDesigner offers (checks, for instance), so maybe it will become a moot point.

The binary file is an image of the ERD. If we're moving away from the DBDesigner solution, maybe this is an option: https://github.com/BurntSushi/erd/blob/master/README.md

If we go with the above, we could probably put together a quick script that would take our text representation of the schema and create a sql file from it.

tmrietz commented on schema/schema_updated_17.sql:
```diff
@@ -0,0 +1,88 @@
+
+
+
+CREATE TABLE "sub-prompt" (
```
I missed this but "sub_prompt" for consistency instead of "sub-prompt"

andrew-rietz replied:
Updated

tmrietz commented:
Ok got it. I think for now let's continue to consider options for the ERD but keep this PR limited to actual schema changes and defining those in new and existing *.sql files in the schema/sql directory. 

You've obviously made most of the changes already on your local DB so just have to make a couple tweaks and write 'em in .sql files for now. The tweaks will address the comments about the "type" field and the "sub-prompt"-> "sub_prompt" change. There's an example at [https://www.postgresql.org/docs/9.4/ddl-constraints.html](url) section 5.3.1 that shows how to add multiple checks for a single field.

andrew-rietz commented:
Moved away from the single sql file approach. Split each table out into its own file and adopted the same syntax / styling you used originally because it's less verbose and easier to follow.
